File: test/e2e/Apps/BasicDotNetIsolated/CallDisabledFunctions.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using Microsoft.Azure.Functions.Worker;
using Microsoft.DurableTask;
using Microsoft.DurableTask.Entities;

namespace Microsoft.Azure.Durable.Tests.E2E;

// Orchestrations that schedule work against the disabled-but-still-deployed functions
// (DisabledActivity / DisabledEntity). These are used by DisabledOrchestrationTests to verify that
// dispatching a disabled function fails the orchestration deterministically instead of poison-looping
// forever. See https://github.com/Azure/azure-functions-durable-extension/issues/3471.
public static class CallDisabledFunctions
{
    [Function(nameof(CallDisabledActivity))]
    public static async Task<string> CallDisabledActivity(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        // DisabledActivity is indexed but has no active listener (its executor is null), so this
        // must surface as a deterministic activity failure rather than hanging forever.
        return await context.CallActivityAsync<string>(nameof(DisabledActivity), "hello");
    }

    [Function(nameof(CallDisabledEntity))]
    public static async Task<string> CallDisabledEntity(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        var entityId = new EntityInstanceId(nameof(DisabledEntity), "disabled-key");

        // DisabledEntity is indexed but has no active listener, so calling an operation on it must
        // fail deterministically rather than poison-looping the entity work item forever.
        await context.Entities.CallEntityAsync(entityId, "someOperation");
        return "should not reach here";
    }
}

File: test/e2e/Apps/BasicNode/src/functions/CallDisabledFunctions.ts
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import * as df from 'durable-functions';
import { OrchestrationContext, OrchestrationHandler } from 'durable-functions';

// Orchestrator that schedules the disabled-but-still-deployed DisabledActivity. Because the activity
// has no active listener, the dispatch must fail the orchestration deterministically instead of
// poison-looping forever. See https://github.com/Azure/azure-functions-durable-extension/issues/3471.
const CallDisabledActivity: OrchestrationHandler = function* (context: OrchestrationContext) {
    return yield context.df.callActivity('DisabledActivity', 'hello');
};

df.app.orchestration('CallDisabledActivity', CallDisabledActivity);

// Companion orchestrator for the entity dispatch path: calling an operation on the disabled-but-still-
// deployed DisabledEntity must fail the orchestration deterministically rather than poison-looping.
const CallDisabledEntity: OrchestrationHandler = function* (context: OrchestrationContext) {
    const entityId = new df.EntityId('DisabledEntity', 'disabled-key');
    yield context.df.callEntity(entityId, 'someOperation');
    return 'should not reach here';
};

df.app.orchestration('CallDisabledEntity', CallDisabledEntity);
